alfresco-search/src/main/java/org/alfresco/repo/search/impl/lucene/analysis/SynonymAwareStopFilterFactory.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.lucene.analysis.StopFilter;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.core.StopFilterFactory;
+import org.apache.lucene.analysis.synonym.SynonymGraphFilter;
+import org.apache.lucene.analysis.tokenattributes.PositionLengthAttribute;
+import org.apache.lucene.analysis.tokenattributes.TypeAttribute;
+
+import java.util.Map;
+
+/**
+ * A {@link StopFilter} which doesn't remove stop tokens previously marked as part of a synonym.
+ * When using a stop filter in a query analyzer which already includes a synonym filter, there are two options:
+ * the filter could be defined before or after the synonym filter.
+ *
+ * However, the first way (before) doesn’t make so much sense, because terms that are stopwords and that are, at the same
+ * time, part of a synonym will be removed before the synonym detection. As consequence of that no synonym detection will happen.
+ *
+ * If we have
+ *
+ * <ul>
+ *     <li>a stopwords list consisting of one term (of)</li>
+ *     <li>and a single synonym definition like "out of warranty,oow"</li>
+ *     <li>a query which should match the synonym: q=out of warranty</li>
+ * </ul>
+ *
+ * a stop filter defined before the synonym filter would remove the "of" term from the query, therefore causing the missing
+ * synonym detection. If we postpone the stop filter after the synonym filter, the synonym will be detected, but the
+ * stop filter later would still remove the "of" token causing the open issue well described in https://issues.apache.org/jira/browse/LUCENE-4065
+ *
+ * Being related with FilteringTokenFilter (the StopFilter superclass) LUCENE-4065 has a wider scope which doesn't affect only
+ * stopwords removal. However, in order to mitigate the issue above, this custom stop filter takes in account the typeAttribute of the
+ * incoming tokens; it removes them only if they are stopwords and if they haven't been marked as part of a synonym ({@link SynonymGraphFilter#TYPE_SYNONYM})
+ *
+ * @see SynonymGraphFilter
+ * @author Andrea Gazzarini
+ */
+public class SynonymAwareStopFilterFactory extends StopFilterFactory
+{
+    class SynonymAwareStopFilter extends StopFilter
+    {
+
+        private TypeAttribute typeAttribute = addAttribute(TypeAttribute.class);
+        private PositionLengthAttribute positionLengthAttribute = addAttribute(PositionLengthAttribute.class);
+
+        private int synonymSpans;
+
+        SynonymAwareStopFilter(TokenStream in, CharArraySet stopwords)
+        {
+            super(in, stopwords);
+        }
+
+        @Override
+        protected boolean accept()
+        {
+            if (isSynonymToken())
+            {
+                synonymSpans = positionLengthAttribute.getPositionLength() > 1
+                        ? positionLengthAttribute.getPositionLength()
+                        : 0;
+                return true;
+            }
+
+            return (--synonymSpans > 0) || super.accept();
```

Comment:
This doesn't look like it will cope with overshadowed multi-word synonyms:
```
Brilliant Company Document for Employees,BCDE
Company Document,CD
```
I think `synonymSpans` will be overwritten before the end of BCDE and the "for" will be a stop word.  Maybe this case isn't supported anyway though?  I was thought that there should be a `max` used when setting synonymSpans.

Reply:
Synonyms overlapping has been/is a huge topic which has been solved in the current implementation by choosing a single "winner" among potential candidates.    

I don't remember exactly what is the logic (I've dig a lot in the SynonymsGraph few months ago, but I'm happy to dig again for providing more info) but it seems the longer synonym wins. In this case, having something like

`"this book is about brilliant company document for employees" `
   
produces the following query:
    
```
title:this 
title:book 
title:is
title:about
title:about (title:bcde PhraseQuery(title:"brilliant company document for employees"))
```    
where you can see the second synonym set (Company Document, CD) is ignored.

Reply:
Ok - that behaviour sounds good. I don't see how it works though. The only `synonymSpans > 0` check comes after `synonymSpans` has been overwritten.

Reply:
Ok, I got you. I'm not sure if this comment is too large for a review comment. However, have a seat :D

```
stopwords = ["for"]
synonyms = ["Brilliant Company Document for Employees,BCDE"]
```

```
q=about brilliant Company document for Employees or for managers
```

Here's the token stream produced by applying a simple analyzer which includes:

- a whitespace tokenizer
- a lowercase filter
- a synonym graph filter
- the custom stop filter

| Term  |Start Offset | End Offset| Position Length| Type|
| ------------- | ------------- | ------------- | ------------- | ------------- |
| about  | 0  | 5 |1|ALPHANUM|
| bcde  |6|46|5|SYNONYM|
|brilliant|6|15|1|ALPHANUM|
|company|16|23|1|ALPHANUM|
|document|24|32|1|ALPHANUM|
|for|33|36|1|ALPHANUM|
|employees|37|46|1|ALPHANUM|
|or|47|49|1|ALPHANUM|
|for|50|54|1|ALPHANUM|
|managers|54|62|1|ALPHANUM|

As you can see the filter is assigning a value to the synonymSpans only when the second token (bcde) arrives, because it has a value > 1. This is saying "Hey, you will find the next 5 token marked as plain alphanum tokens, but they are part of a synonym". 
For that reason, each time a new token (e.g brilliant, company,...) will flow through the accept method, it will decrement that span; until that value will be greater than 0 we are sure we are dealing with a token which is part of a previously detected synonym; as consequence of that it won't be removed, even if that is a stopword. 
Going ahead with the example, ater the "employees" token that value will be 0 and that means the subsequent "for" token will be removed because of the plain stopfilter behaviour.

That will produce the following query:

```
title:about (title:bcde PhraseQuery(title:"brilliant company document for employees")) title:or title:managers
```

Make sense?

Reply:
Yes - thank you. This was helpful! :+1: